xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java

```diff
@@ -422,6 +448,7 @@ private void reportSuccess() {
   /** Shutdown this instance and release resources. */
   public void shutdown() {
     logger.log(Level.FINER, "Shutdown");
+    releaseOldSuppliers(curListener.get());
```

Comment:
`releaseOldSuppliers` is not thread safe. I think you should do this after `xdsClient.cancelLdsResourceWatch()`.

Reply:
Good point about canceling the LdsResourceWatch. Done.

About releaseOldSuppliers() thread-safety - it gets the old value atomically and "cleans it up" without using any other variables. And the old value is not used anywhere else - only used by this instance.

Reply:
> About releaseOldSuppliers() thread-safety - it gets the old value atomically and "cleans it up" without using any other variables. And the old value is not used anywhere else - only used by this instance.

Oh, that's right.